### PR TITLE
Optimisation of the MULMOD opcode

### PR DESCRIPTION
**Pull request by vkomenda**

### CODE REVIEW

**Comment by vkomenda** [comment]
## Benchmark comparison

### Old version

```
mulmod randomly generated ints                                                                             
                        time:   [8.3800 us 8.3820 us 8.3842 us]
Found 17 outliers among 100 measurements (17.00%)
  12 (12.00%) high mild
  5 (5.00%) high severe

mulmod 500 times        time:   [6.8972 ms 6.8981 ms 6.8990 ms]                              
                        change: [+884.24% +884.69% +885.02%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

mulmod 1000 times       time:   [13.809 ms 13.810 ms 13.812 ms]                               
                        change: [+885.87% +886.44% +886.94%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe
```

### New version

```
mulmod randomly generated ints                                                                             
                        time:   [2.7198 us 2.7212 us 2.7228 us]
                        change: [-67.555% -67.515% -67.470%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe

mulmod 500 times        time:   [741.21 us 741.46 us 741.77 us]                             
                        change: [-89.254% -89.249% -89.243%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) high mild
  7 (7.00%) high severe

mulmod 1000 times       time:   [1.4717 ms 1.4722 ms 1.4727 ms]                               
                        change: [-89.345% -89.341% -89.336%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe
```

Thus we have a 884% improvement. When we remove conversion of `uint` types to `rug` and back again, the speed can increase yet again but not by that much.

**Comment by varasev** [comment]
Could you clarify how could I run the benchmarking tests? Is there some CLI option for that?

**Comment by varasev** [comment]
I'd also propose to add more benchmarks here (and compare them with the old version):

- `mulmod(U256::MAX, U256::MAX, 2)`
- `mulmod(U256::MAX, U256::MAX, 5)`
- `mulmod(U256::MAX, U256::MAX, 10)`
- `mulmod(U256::MAX, U256::MAX, 0xFFFFFFFFFFFFFFFFF)` (division by some big number)

and would compare that case with the random ints for 500 and 1000 iterations:

```
mulmod(
  0xcc2de1f8ec6cc9a24ed2c48b856637f9e45f0a5feee21a196aa42a290ef454ca, 
  0xb8e0a2b6b1587398c28bf9e9d34ea24ba34df308eec2acedca363b2fce2c25db, 
  0x5ed6db9489224124a1a4110ec8bec8b01369c8b549a4b8c4388a1796dc35a937
)
```

**Comment by afck** [comment]
My suspicion is that the build error is because with `rug`, the right operand of `<` does not have to be `u64` anymore if the left one is, because it now can also be an `Integer`. So the `step.into()` needs to be replaced by `u64::from(step)` in some places in the code.

File: ethcore/evm/src/interpreter/mod.rs [hunk 1]
```diff
@@ -1156,7 +1160,6 @@ fn get_and_reset_sign(value: U256) -> (U256, bool) {
 	let sign = arr[3].leading_zeros() == 0;
 	(set_sign(value, sign), sign)
 }
-
```
[the remark this answers is not in the document]
Comment (vkomenda) [reply ethcore/evm/src/interpreter/mod.rs]: Do you mean the newline?

**Comment by vkomenda** [comment]
New results together with additional single-word modulus cases for 5 and 11 (`cargo bench` in `ethcore/evm`):
```
mulmod randomly generated ints, 500 times                                                                            
                        time:   [773.13 us 773.45 us 773.84 us]
                        change: [-6.1069% -6.0188% -5.9305%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 18 outliers among 100 measurements (18.00%)
  4 (4.00%) high mild
  14 (14.00%) high severe

mulmod randomly generated ints, 1000 times                                                                             
                        time:   [1.5461 ms 1.5467 ms 1.5474 ms]
                        change: [-6.7707% -6.7003% -6.6302%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  3 (3.00%) high mild
  9 (9.00%) high severe

mulmod modulo 1, 500 times                                                                            
                        time:   [680.70 us 681.11 us 681.54 us]
                        change: [-3.7842% -3.6330% -3.5047%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  5 (5.00%) high mild
  6 (6.00%) high severe

mulmod modulo 1, 1000 times                                                                             
                        time:   [1.3389 ms 1.3395 ms 1.3402 ms]
                        change: [-5.2665% -5.1733% -5.0786%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  5 (5.00%) high severe

mulmod modulo 5, 500 times                                                                            
                        time:   [669.32 us 670.06 us 671.44 us]
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe

mulmod modulo 5, 1000 times                                                                             
                        time:   [1.3365 ms 1.3369 ms 1.3374 ms]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe

mulmod modulo 11, 500 times                                                                            
                        time:   [699.90 us 700.19 us 700.56 us]
Found 16 outliers among 100 measurements (16.00%)
  4 (4.00%) low severe
  5 (5.00%) low mild
  1 (1.00%) high mild
  6 (6.00%) high severe

mulmod modulo 11, 1000 times                                                                             
                        time:   [1.3978 ms 1.3987 ms 1.3999 ms]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
```
As @afck pointed out, the deciding factor is that we benchmark on single-word moduli, which makes it run faster. Even though the full-size modulus case is only about 10% slower.

**Comment by varasev** [comment]
> New results together with additional single-word modulus cases for 5 and 11...

Can you execute the same tests for the old version?

File: ethcore/evm/benches/basic.rs [hunk 2]
```diff
@@ -157,3 +166,104 @@ fn result(r: Result<evm::GasLeft>) -> U256 {
...
+}
+
+/// Compute mulmod(U256::MAX, U256::MAX, 5) 1000 times.
+fn mulmod5_1000(b: &mut Criterion) {
+	b.bench_function("mulmod modulo 5, 1000 times", |b| {
+		run_code(b, "6103e85b6001900360057fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff80095080600357".from_hex().unwrap());
+	});
+}
+
+/// Compute mulmod(U256::MAX, U256::MAX, 1) 500 times.
```
Comment (varasev) [inline ethcore/evm/benches/basic.rs]: The comment here should be `/// Compute mulmod(U256::MAX, U256::MAX, 11) 500 times.`

File: ethcore/evm/benches/basic.rs [hunk 3]
```diff
@@ -157,3 +166,104 @@ fn result(r: Result<evm::GasLeft>) -> U256 {
...
+fn mulmod11_500(b: &mut Criterion) {
+	b.bench_function("mulmod modulo 11, 500 times", |b| {
+		run_code(b, "6101f45b60019003600b7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff80095080600357".from_hex().unwrap());
+	});
+}
+
+/// Compute mulmod(U256::MAX, U256::MAX, 1) 1000 times.
```
Comment (varasev) [inline ethcore/evm/benches/basic.rs]: The comment here should be `/// Compute mulmod(U256::MAX, U256::MAX, 11) 1000 times.`

**Comment by vkomenda** [comment]
## The updated benchmark comparison

The improvement is 78% to 90%.

### Old

```
mulmod randomly generated ints, 500 times                                                                             
                        time:   [3.7005 ms 3.7017 ms 3.7032 ms]
                        change: [+370.32% +370.57% +370.83%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 15 outliers among 100 measurements (15.00%)
  2 (2.00%) high mild
  13 (13.00%) high severe

mulmod randomly generated ints, 1000 times                                                                             
                        time:   [7.3744 ms 7.3759 ms 7.3774 ms]
                        change: [+368.02% +368.44% +368.79%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe

mulmod modulo 1, 500 times                                                                             
                        time:   [6.9113 ms 6.9126 ms 6.9142 ms]
                        change: [+908.05% +908.67% +909.26%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 9 outliers among 100 measurements (9.00%)
  4 (4.00%) high mild
  5 (5.00%) high severe

mulmod modulo 1, 1000 times                                                                             
                        time:   [13.822 ms 13.825 ms 13.830 ms]
                        change: [+918.43% +919.10% +919.71%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  3 (3.00%) high severe

mulmod modulo 5, 500 times                                                                             
                        time:   [6.9462 ms 6.9466 ms 6.9471 ms]
                        change: [+919.14% +919.72% +920.16%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe

mulmod modulo 5, 1000 times                                                                             
                        time:   [13.898 ms 13.899 ms 13.900 ms]
                        change: [+912.25% +913.16% +913.79%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) high mild
  5 (5.00%) high severe

mulmod modulo 11, 500 times                                                                             
                        time:   [6.9486 ms 6.9497 ms 6.9510 ms]
                        change: [+879.20% +879.92% +880.55%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 12 outliers among 100 measurements (12.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  7 (7.00%) high severe

mulmod modulo 11, 1000 times                                                                             
                        time:   [13.903 ms 13.906 ms 13.909 ms]
                        change: [+880.37% +880.93% +881.42%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe

mulmod modulo random 256-bit number, 500 times                                                                             
                        time:   [3.7789 ms 3.7800 ms 3.7813 ms]
                        change: [+380.56% +380.82% +381.08%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 15 outliers among 100 measurements (15.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  12 (12.00%) high severe

mulmod modulo random 256-bit number, 1000 times                                                                             
                        time:   [7.5611 ms 7.5625 ms 7.5640 ms]
                        change: [+377.07% +377.36% +377.62%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 14 outliers among 100 measurements (14.00%)
  8 (8.00%) high mild
  6 (6.00%) high severe
```

### New

```
mulmod randomly generated ints, 500 times                                                                            
                        time:   [805.80 us 807.37 us 809.88 us]
                        change: [-78.229% -78.204% -78.175%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

mulmod randomly generated ints, 1000 times                                                                             
                        time:   [1.6069 ms 1.6073 ms 1.6077 ms]
                        change: [-78.207% -78.197% -78.187%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe

mulmod modulo 1, 500 times                                                                            
                        time:   [701.89 us 702.29 us 702.73 us]
                        change: [-89.848% -89.842% -89.834%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

mulmod modulo 1, 1000 times                                                                             
                        time:   [1.3869 ms 1.3875 ms 1.3882 ms]
                        change: [-89.970% -89.964% -89.958%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  2 (2.00%) low mild
  4 (4.00%) high mild
  7 (7.00%) high severe

mulmod modulo 5, 500 times                                                                            
                        time:   [687.15 us 687.51 us 687.93 us]
                        change: [-90.110% -90.105% -90.099%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  3 (3.00%) low mild
  3 (3.00%) high mild
  7 (7.00%) high severe

mulmod modulo 5, 1000 times                                                                             
                        time:   [1.3890 ms 1.3897 ms 1.3904 ms]
                        change: [-90.005% -89.998% -89.988%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe

mulmod modulo 11, 500 times                                                                            
                        time:   [710.75 us 710.92 us 711.10 us]
                        change: [-89.774% -89.767% -89.756%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  4 (4.00%) high severe

mulmod modulo 11, 1000 times                                                                             
                        time:   [1.4260 ms 1.4266 ms 1.4273 ms]
                        change: [-89.750% -89.744% -89.736%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) low mild
  1 (1.00%) high mild
  5 (5.00%) high severe

mulmod modulo random 256-bit number, 500 times                                                                            
                        time:   [806.39 us 806.51 us 806.66 us]
                        change: [-78.669% -78.655% -78.639%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 15 outliers among 100 measurements (15.00%)
  1 (1.00%) low mild
  5 (5.00%) high mild
  9 (9.00%) high severe

mulmod modulo random 256-bit number, 1000 times                                                                             
                        time:   [1.6114 ms 1.6119 ms 1.6126 ms]
                        change: [-78.699% -78.687% -78.675%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  5 (5.00%) high severe
```

**Comment by varasev** [comment]
> ## The updated benchmark comparison
> The improvement is 78% to 90%.

I see the next picture: https://docs.google.com/spreadsheets/d/10imlaMVaqTPmh0rvnGbFfu2FnEG46DO4M0vFatchaew/edit?usp=sharing

File: ethcore/evm/src/interpreter/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (afck) [reply ethcore/evm/src/interpreter/mod.rs]: Yes: Between functions there should usually be newlines.

**Pull request merged**